scripts/updateModHistory.php

```diff
@@ -0,0 +1,219 @@
+git diff --name-only $commonAncestorHash $head | while read -r filename; do
+  echo "filename:"
+  echo "\$filename"
+  echo "modified:"
+  echo "$(git log -1 --format='%aI' -- \$filename)"
+  echo "contributors:"
+  git log --format='%an' -- \$filename|awk '!a[$0]++'
+done
+COMMAND;
+
+echo timeStamp() . " - Retrieving commit authors and last commit date/time of modified files... \n";
+
+$fileCounter = 0;
+$modifiedFiles = [];
+
+$proc = popen($modifiedFilescommand, 'rb');
+while (($line = fgets($proc)) !== false) {
+    processGitDiffLine(rtrim($line, "\n\r"), $modifiedFiles);
+    if (! $runningInGithubActions) {
+        fwrite(
+            STDERR,
+            sprintf("\033[0G{$fileCounter} of {$numOfFilesWithDiff} files read...", "", "")
+        );
+    }
+}
+pclose($proc);
+
+echo " done\n";
+
+$s = ($modifiedFiles > 2) ? "s" : "";
+echo timeStamp() . " - Retrieved author$s and last commit date$s/time$s for " . (count($modifiedFiles) - 1) . " file$s\n";
+if (count($modifiedFiles) === 1) {
+    // there will always be at least 1 entry with the last commit hash
+    exit(1);
+}
+
+$mergedModHistory = array_merge($modHistoryArray, $modifiedFiles);
+
+echo timeStamp() . " - Writing modification history file... ";
+
+$fp = fopen($modHistoryFile, "w");
+fwrite($fp, "<?php\n\n/* This is a generated file */\n\nreturn [\n");
+foreach ($mergedModHistory as $fileName => $fileProps) {
+    if ($fileName === "last commit hash") {
+        fwrite($fp, "    \"last commit hash\" => \"" . implode("", $fileProps) . "\",\n");
+        continue;
+    }
+    $newModHistoryString = '    "' . $fileName . "\" => [\n";
+    $newModHistoryString .= "        \"modified\" => \"" . ($fileProps["modified"] ?? "") . "\",\n";
+    $newModHistoryString .= "        \"contributors\" => [\n";
+    if (isset($fileProps["contributors"])) {
+        if (!is_array($fileProps["contributors"])) {
+            exit("Non-array contributors list\n");
+        }
+        foreach ($fileProps["contributors"] as $contributor) {
+            $newModHistoryString .= "            \"" . $contributor . "\",\n";
+        }
+    }
+    $newModHistoryString .= "        ],\n";
+    $newModHistoryString .= "    ],\n";
+    fwrite($fp, $newModHistoryString);
+}
+fwrite($fp, "];\n");
+fclose($fp);
+
+echo "done at " . date('H:i:s') . "\n";
+
+function timeStamp(): string {
+    return "[" . date('H:i:s') . "]";
+}
+
+function processGitDiffLine($line, &$modifiedFiles): void {
+    static $currentType = "";
+    static $currentFile = "";
+    global $fileCounter;
+
+    switch ($line) {
+        case "filename:":
+            $currentType = "filename";
+            $fileCounter++;
+            return;
+        case "modified:":
+            $currentType = "modDateTime";
+            return;
+        case "contributors:":
+            $currentType = "contributors";
+            return;
+        case "last commit hash:":
+            $currentType = "commitHash";
+            return;
+    }
```

Comment:
Maybe use `match` here? Or is it expected to not be one of these cases?

Reply:
It should only be one of these cases and everything else is skipped by the function.

Reply:
I just realized that I was wrong: this should be a `switch` so the function can return as `$line` won't be any of the above cases most of the time but will contain all modification dates/times and the names of every commit author.  Ie. when `$currentType` is set but `$line` is not any of those cases, it contains the information we are actually looking for (filenames, dates/times, commit authors).